File: app/src/main/java/com/tammamkhalaf/myuaeguide/User/UserDashboard.java
package com.tammamkhalaf.myuaeguide.User;

import android.content.Intent;
import android.graphics.drawable.GradientDrawable;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.navigation.NavigationView;
import com.tammamkhalaf.myuaeguide.Categories.Hotels.HotelService;
import com.tammamkhalaf.myuaeguide.Categories.Hotels.HotelServiceBuilder;
import com.tammamkhalaf.myuaeguide.Categories.Hotels.searchHotelRegionOne.Properties;
import com.tammamkhalaf.myuaeguide.Categories.Hotels.searchHotelRegionOne.Suggestion;
import com.tammamkhalaf.myuaeguide.Categories.NearbyPlaces.Results;
import com.tammamkhalaf.myuaeguide.Categories.NearbyPlaces.Place;
import com.tammamkhalaf.myuaeguide.Categories.NearbyPlaces.ServiceBuilder;
import com.tammamkhalaf.myuaeguide.Categories.NearbyPlaces.TrueWayPlacesService;
import com.tammamkhalaf.myuaeguide.Common.LoginSignup.RetailerStartUpScreen;
import com.tammamkhalaf.myuaeguide.HelperClasses.HomeAdapter.Categories.CategoriesAdapter;
import com.tammamkhalaf.myuaeguide.HelperClasses.HomeAdapter.Categories.CategoriesHelperClass;
import com.tammamkhalaf.myuaeguide.HelperClasses.HomeAdapter.Featured.FeaturedAdapter;
import com.tammamkhalaf.myuaeguide.HelperClasses.HomeAdapter.Featured.FeaturedHelperClass;
import com.tammamkhalaf.myuaeguide.HelperClasses.HomeAdapter.MostViewed.MostViewedAdapter;
import com.tammamkhalaf.myuaeguide.HelperClasses.HomeAdapter.MostViewed.MostViewedHelperClass;
import com.tammamkhalaf.myuaeguide.R;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;


public class UserDashboard extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {

    private static final String TAG = "UserDashboard";
    RecyclerView featuredRecycler, mostViewedRecycler, categoriesRecycler;
    RecyclerView.Adapter adapter;
    ImageView menuIcon;

    LinearLayout content;
    static final float END_SCALE = 0.7f;
    DrawerLayout drawerLayout;
    NavigationView navigationView;
    Call<Results> call;
    ArrayList<Place> places;
    ArrayList<Suggestion> suggestions;

    //todo add section for the mahrajanat in uae mahrajan zayed etc ....

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_user_dashboard);


        //Hooks
        featuredRecycler = findViewById(R.id.featured_recycler);
        mostViewedRecycler = findViewById(R.id.most_viewed_recycler);
        categoriesRecycler = findViewById(R.id.categories_recycler);

        drawerLayout = findViewById(R.id.drawer_layout);
        navigationView = findViewById(R.id.navigation_view);

        menuIcon = findViewById(R.id.menu_icon);

        content = findViewById(R.id.content);


        //Functions will be executed automatically when this activity will be created
        featuredRecycler();
        mostViewedRecycler();
        categoriesRecycler();
        navigationDrawer();
    }

    //region navigation drawer

    private void navigationDrawer() {
        navigationView.bringToFront();
        navigationView.setNavigationItemSelectedListener(this);
        navigationView.setCheckedItem(R.id.nav_home);

        menuIcon.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if (drawerLayout.isDrawerVisible(GravityCompat.START)) {
                    drawerLayout.closeDrawer(GravityCompat.START);
                } else {
                    drawerLayout.openDrawer(GravityCompat.START);
                }
            }
        });
        animateNavigationDrawer();
    }

    private void animateNavigationDrawer() {

        //Add any color or remove it to use the default one!
        //To make it transparent use Color.Transparent in side setScrimColor();
        drawerLayout.setScrimColor(getResources().getColor(R.color.banner_background_light));//Color.TRANSPARENT
        drawerLayout.addDrawerListener(new DrawerLayout.SimpleDrawerListener() {
            @Override
            public void onDrawerSlide(View drawerView, float slideOffset) {

                // Scale the View based on current slide offset
                final float diffScaledOffset = slideOffset * (1 - END_SCALE);
                final float offsetScale = 1 - diffScaledOffset;
                content.setScaleX(offsetScale);
                content.setScaleY(offsetScale);

                // Translate the View, accounting for the scaled width
                final float xOffset = drawerView.getWidth() * slideOffset;
                final float xOffsetDiff = content.getWidth() * diffScaledOffset / 2;
                final float xTranslation = xOffset - xOffsetDiff;
                content.setTranslationX(xTranslation);
            }
        });

    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.nav_all_categories) {
            startActivity(new Intent(getApplicationContext(), AllCategories.class));
        } else {
            Toast.makeText(this, "Under Development", Toast.LENGTH_SHORT).show();
        }

        return true;
    }

    @Override
    public void onBackPressed() {
        if (drawerLayout.isDrawerVisible(GravityCompat.START)) {
            drawerLayout.closeDrawer(GravityCompat.START);
        } else
            super.onBackPressed();
    }
    //endregion

    //region featured
    private void featuredRecycler() {

        featuredRecycler.setHasFixedSize(true);
        featuredRecycler.setLayoutManager(new LinearLayoutManager(this, RecyclerView.HORIZONTAL, false));

        ArrayList<FeaturedHelperClass> featuredLocations = new ArrayList<>();

        TrueWayPlacesService service = ServiceBuilder.buildService(TrueWayPlacesService.class);

        int[] images = {R.drawable.sheikh_zayed_grand_mosque, R.drawable.burj_al_arab};
        // todo add array of images

        // todo 429 Too Many Requests
        //  todo use Rxjava to separate time of consuming api

        HashMap<String, String> TrueWayPlacesFilterMap = new HashMap<>();

        String locationCoordinates = "24.199168,55.719391";

        TrueWayPlacesFilterMap.put("location", locationCoordinates);
        /**
         * type >>>>
         airport,amusement_park,aquarium,art_gallery,atm,
         bakery,bank,bar,beauty_salon,bicycle_store,book_store,bowling,bus_station
         cafe,campground,car_dealer,car_rental,car_repair,car_wash,casino,cemetery,church,cinema,city_hall,clothing_store,convenience_store,courthouse
         dentist,department_store,doctor,electrician,electronics_store,embassy
         fire_station,flowers_store,funeral_service,furniture_store,gas_station
         government_office,grocery_store,gym,
         hairdressing_salon,hardware_store,homegoodsstore,hospital,insurance_agency,jewelry_store
         laundry,lawyer,library,liquor_store,locksmith,lodging,mosque,museum,night_club
         park,parking,pet_store,pharmacy,plumber,police_station,post_office,primary_school,
         rail_station,realestateagency,restaurant,rv_park
         school,secondary_school,shoe_store,shopping_center,spa,stadium,storage,store,subway_station,supermarket,synagogue
         taxi_stand,temple,tourist_attraction,train_station,transit_station,travel_agency,university
         veterinarian,zoo
         * */
        TrueWayPlacesFilterMap.put("type", "cafe");

        call = service.findPlacesNearby(TrueWayPlacesFilterMap, 10000, "en");

        call.enqueue(new Callback<Results>() {
            @Override
            public void onResponse(Call<Results> call, Response<Results> response) {
                if (response.isSuccessful()) {
                    places = (ArrayList<Place>) response.body().getResults();
                    for (Place place : places) {
                        featuredLocations.add(new FeaturedHelperClass(R.drawable.coffee_shop, place.getName(), place.getPhoneNumber()));
                    }
                    adapter = new FeaturedAdapter(featuredLocations, UserDashboard.this);
                    featuredRecycler.setAdapter(adapter);
                } else if (response.code() == 401) {
                    Toast.makeText(UserDashboard.this, R.string.SessionExpired, Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(UserDashboard.this, R.string.FailedToRetrieveItems, Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Results> call, Throwable t) {
                if (t instanceof IOException) {
                    Toast.makeText(UserDashboard.this, R.string.AConnectionErrorOccurred, Toast.LENGTH_SHORT).show();
                } else {
                    Toast.makeText(UserDashboard.this, R.string.FailedToRetrieveItems, Toast.LENGTH_SHORT).show();
                }
            }
        });

    }
    //endregion

    //region categories
    private void categoriesRecycler() {

        //All Gradients
        GradientDrawable gradient2 = new GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, new int[]{0xffd4cbe5, 0xffd4cbe5});
        GradientDrawable gradient1 = new GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, new int[]{0xff7adccf, 0xff7adccf});
        GradientDrawable gradient3 = new GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, new int[]{0xfff7c59f, 0xFFf7c59f});
        GradientDrawable gradient4 = new GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, new int[]{0xffb8d7f5, 0xffb8d7f5});


        ArrayList<CategoriesHelperClass> categoriesHelperClasses = new ArrayList<>();
        categoriesHelperClasses.add(new CategoriesHelperClass(gradient1, R.drawable.school_cat, getString(R.string.education)));
        categoriesHelperClasses.add(new CategoriesHelperClass(gradient2, R.drawable.hospital_cat, getString(R.string.hospital)));
        categoriesHelperClasses.add(new CategoriesHelperClass(gradient3, R.drawable.res_cat, getString(R.string.restaurant)));
        categoriesHelperClasses.add(new CategoriesHelperClass(gradient4, R.drawable.shopping_cat, getString(R.string.shopping)));
        categoriesHelperClasses.add(new CategoriesHelperClass(gradient1, R.drawable.transport_cat, getString(R.string.transport)));


        categoriesRecycler.setHasFixedSize(true);
        adapter = new CategoriesAdapter(categoriesHelperClasses);
        categoriesRecycler.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));
        categoriesRecycler.setAdapter(adapter);

    }
    //endregion

    //region mostViewed
    private void mostViewedRecycler() {

        mostViewedRecycler.setHasFixedSize(true);
        mostViewedRecycler.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false));

        HotelService service = HotelServiceBuilder.buildService(HotelService.class);

        Call<Properties> call = service.searchHotel("dubai", "en_US");

        ArrayList<MostViewedHelperClass> mostViewedLocations = new ArrayList<>();


        call.enqueue(new Callback<Properties>() {
            @Override
            public void onResponse(Call<Properties> call, Response<Properties> response) {
                if (response.isSuccessful()) {
                    suggestions = (ArrayList<Suggestion>) response.body().suggestions;
                    Log.d(TAG, "onResponse: suggestions" + suggestions);
                    for (Suggestion suggestion : suggestions) {
                        for (int i = 0; i < suggestion.entities.size(); i++) {
                            mostViewedLocations.add(new MostViewedHelperClass(R.drawable.beds,
                                    suggestion.entities.get(i).name,
                                    suggestion.group));
                        }
                    }
                    adapter = new MostViewedAdapter(mostViewedLocations);
                    mostViewedRecycler.setAdapter(adapter);
                } else if (response.code() == 401) {
                    Toast.makeText(UserDashboard.this, R.string.SessionExpired, Toast.LENGTH_SHORT).show();
                } else {
                    Log.d(TAG, "onResponse: Not-Successful Response Code = " + response.code() +
                            "-->>> Response Error Body:" + response.errorBody().toString());
                    Toast.makeText(UserDashboard.this, R.string.FailedToRetrieveItems, Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<Properties> call, Throwable t) {
                if (t instanceof IOException) {
                    Toast.makeText(UserDashboard.this, R.string.AConnectionErrorOccurred, Toast.LENGTH_SHORT).show();
                    Log.d(TAG, "onFailure: " + t.getLocalizedMessage());
                } else {
                    Log.d(TAG, "onFailure: " + t.getLocalizedMessage());
                    Toast.makeText(UserDashboard.this, R.string.FailedToRetrieveItems, Toast.LENGTH_SHORT).show();
                }
            }
        });


    }
    //endregion

    public void callRetailerScreen(View view) {
        startActivity(new Intent(getApplicationContext(), RetailerStartUpScreen.class));
    }

}

File: app/src/main/java/com/tammamkhalaf/myuaeguide/user/UserDashboard.kt
package com.tammamkhalaf.myuaeguide.user

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.GradientDrawable
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.MenuItem
import android.view.View
import android.view.View.GONE
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.*
import androidx.annotation.NonNull
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.drawerlayout.widget.DrawerLayout.SimpleDrawerListener
import androidx.lifecycle.*
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.facebook.shimmer.ShimmerFrameLayout
import com.google.android.gms.common.api.internal.BackgroundDetector.initialize
import com.google.android.gms.location.*
import com.google.android.material.navigation.NavigationView
import com.nostra13.universalimageloader.core.ImageLoader
import com.tammamkhalaf.myuaeguide.R
import com.tammamkhalaf.myuaeguide.R.string
import com.tammamkhalaf.myuaeguide.chat.ChatActivity
import com.tammamkhalaf.myuaeguide.common.loginSignup.RetailerStartUpScreen
import com.tammamkhalaf.myuaeguide.common.loginSignup.login.java.PhoneAuthActivity
import com.tammamkhalaf.myuaeguide.databases.firebase.storage.Utility.UniversalImageLoader
import com.tammamkhalaf.myuaeguide.helperClasses.homeAdapter.categories.CategoriesAdapter
import com.tammamkhalaf.myuaeguide.helperClasses.homeAdapter.categories.CategoriesHelperClass
import com.tammamkhalaf.myuaeguide.helperClasses.homeAdapter.featured.FeaturedAdapter
import com.tammamkhalaf.myuaeguide.helperClasses.homeAdapter.featured.FeaturedHelperClass
import com.tammamkhalaf.myuaeguide.helperClasses.homeAdapter.mostViewed.MostViewedAdapter
import com.tammamkhalaf.myuaeguide.helperClasses.homeAdapter.mostViewed.MostViewedHelperClass
import com.tammamkhalaf.myuaeguide.locationOwner.SettingsActivity
import com.tammamkhalaf.myuaeguide.viewmodels.UserDashboardViewModel
import dagger.hilt.android.AndroidEntryPoint
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.ObservableEmitter
import io.reactivex.rxjava3.schedulers.Schedulers
import org.reactivestreams.Subscriber
import java.util.*
import java.util.concurrent.TimeUnit


@AndroidEntryPoint
open class UserDashboard : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    var featuredRecycler: RecyclerView? = null

    var mostViewedRecycler: RecyclerView? = null

    var categoriesRecycler: RecyclerView? = null
    var categoriesAdapter: CategoriesAdapter? = null

    var menuIcon: ImageView? = null
    var content: LinearLayout? = null

    var drawerLayout: DrawerLayout? = null
    var navigationView: NavigationView? = null

    lateinit var latitude:String
    lateinit var longitude:String

    private lateinit var viewModel: UserDashboardViewModel

    private var listOfMostViewedAdapter = ArrayList<MostViewedHelperClass>()
    private var listOfFeaturedAdapter = ArrayList<FeaturedHelperClass>()

    lateinit var container: ShimmerFrameLayout

    /**
     * permissions request code
     */
    private val REQUEST_CODE_ASK_PERMISSIONS = 1

    /**
     * Permissions that need to be explicitly requested from end user.
     */
    private val REQUIRED_SDK_PERMISSIONS = arrayOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.ACCESS_COARSE_LOCATION)


    //Declaring the needed Variables
    lateinit var fusedLocationProviderClient: FusedLocationProviderClient
    lateinit var locationRequest: LocationRequest

    private val MY_PERMISSIONS_REQUEST_LOCATION = 99


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_dashboard)
        window.setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN)

        viewModel = ViewModelProvider(this).get(UserDashboardViewModel::class.java)

        //Hooks
        featuredRecycler = findViewById(R.id.featured_recycler)
        mostViewedRecycler = findViewById(R.id.most_viewed_recycler)
        categoriesRecycler = findViewById(R.id.categories_recycler)
        drawerLayout = findViewById(R.id.drawer_layout)
        navigationView = findViewById(R.id.navigation_view)
        menuIcon = findViewById(R.id.menu_icon)
        content = findViewById(R.id.content)

        // Get a reference to the AutoCompleteTextView in the layout
        val textView = findViewById<AutoCompleteTextView>(R.id.autocomplete_suggest)
        // Get the string array
        val countries: Array<out String> = resources.getStringArray(R.array.searchDiscoverExploreSuggest)
        // Create the adapter and set it to the AutoCompleteTextView
        ArrayAdapter<String>(this, android.R.layout.simple_list_item_1, countries).also { adapter ->
            textView.setAdapter(adapter)
        }

        checkPermissions()

        Observable.create { emitter: ObservableEmitter<Any?>? ->
            textView.addTextChangedListener(object : TextWatcher {
                override fun afterTextChanged(s: Editable?) {
                    Log.d(TAG, "afterTextChanged: ${s.toString()}")

                    if (s?.length != 0)
                        emitter?.onNext(s.toString())
                }

                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                    Log.d(TAG, "beforeTextChanged: ${s.toString()}")
                }

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                    Log.d(TAG, "beforeTextChanged: ${s.toString()}")
                }
            })

        }.doOnNext { c -> Log.d(TAG, "here > upstream: $c") }
                .map {
                    //if you need to apply any thing to
                    // object receive it before send it to final step
                    it.toString().trim()
                }.debounce(5, TimeUnit.SECONDS)
                .distinctUntilChanged()
                .filter {
                    (it.toString() != "abc")
                    /*filtering some word based on text or size
                     or other thing you need to filter it here */
                }
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe { d ->
                    Log.d(TAG, "downstream :  --> CALLING here developer API $d")
                    var list = ArrayList<String>()
                    list.add(d.toString())
                    sendDataToApiDemo(list)
                }


        textView.setOnEditorActionListener(TextView.OnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_SEARCH) {
                val text = v.text.toString()
                Toast.makeText(this@UserDashboard, text, Toast.LENGTH_SHORT).show()
                return@OnEditorActionListener true
            }
            false
        })

        //endregion

        //Functions will be executed automatically when this activity will be created
        navigationDrawer()

        categoriesRecycler()

        textView.afterTextChanged {
            /*doSomethingWithText(it)*/
            Log.d(TAG, "onCreate: afterTextChanged I am calling api")
        }


        initImageLoader()
    }

    /**
     * init universal image loader
     */
    private fun initImageLoader() {
        var imageLoader = UniversalImageLoader(this)
        ImageLoader.getInstance().init(imageLoader.config)
    }

    private fun sendDataToApiDemo(data: ArrayList<String>) {

        container.startShimmer()

        viewModel.discoverExplorePlacesHereDeveloper("dmLgAQo631UJfwF5R2hH", "391hkRjz5Z3Ee1h3wz6Kng",
                "$latitude,$longitude", data)

        viewModel.discoverExplorePlacesHereDeveloperLiveData.observe(this,
                Observer {
                    for (item in it.results.items) {
                        var str: StringBuilder
                        if (item.title.length > 21) {
                            str = java.lang.StringBuilder(item.title)
                            str.insert(21, "\n").toString()
                        } else {
                            str = java.lang.StringBuilder(item.title)
                        }
                        listOfMostViewedAdapter.add(MostViewedHelperClass(
                                item.icon,
                                str.toString(),
                                item?.alternativeNames?.get(0)?.name ?: "",
                                item.category.title ?: "Category?",
                                item.openingHours?.label ?: "Opening Hours",
                                item.openingHours?.text?.replace("<br/>", "\n")
                                        ?: "Not Available?",
                                rating = item.averageRating ?: 4.0,
                                id = item.id
                        ))
                    }

                    if (mostViewedRecycler?.adapter != null) // it works second time and later
                        mostViewedRecycler?.adapter?.notifyDataSetChanged()
                    else {
                        mostViewedRecycler?.adapter = MostViewedAdapter(listOfMostViewedAdapter,
                                this)
                    }
                    container.stopShimmer()
                    container.visibility = GONE
                    mostViewedRecycler?.visibility = View.VISIBLE
                })
    }

    abstract class NYTSubscriber<T> : Subscriber<T> {
        fun onCompleted() {}
        override fun onError(e: Throwable?) {
            Log.e(TAG, "onError: ", e)
        }
    }

    private fun EditText.afterTextChanged(afterTextChanged: (String) -> Unit) {
        this.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(editable: Editable?) {
                afterTextChanged.invoke(editable.toString())
                Log.d(TAG, "afterTextChanged: from extension function ${editable.toString()}")
            }
        })
    }

    //region navigation drawer
    private fun navigationDrawer() {
        navigationView!!.bringToFront()
        navigationView!!.setNavigationItemSelectedListener(this)
        navigationView!!.setCheckedItem(R.id.nav_home)
        menuIcon!!.setOnClickListener {
            if (drawerLayout!!.isDrawerVisible(GravityCompat.START)) {
                drawerLayout!!.closeDrawer(GravityCompat.START)
            } else {
                drawerLayout!!.openDrawer(GravityCompat.START)
            }
        }
        animateNavigationDrawer()
    }

    private fun animateNavigationDrawer() {

        //Add any color or remove it to use the default one!
        //To make it transparent use Color.Transparent in side setScrimColor();
        drawerLayout!!.setScrimColor(resources.getColor(R.color.banner_background_light)) //Color.TRANSPARENT
        drawerLayout!!.addDrawerListener(object : SimpleDrawerListener() {
            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {

                // Scale the View based on current slide offset
                val diffScaledOffset = slideOffset * (1 - END_SCALE)
                val offsetScale = 1 - diffScaledOffset
                content!!.scaleX = offsetScale
                content!!.scaleY = offsetScale

                // Translate the View, accounting for the scaled width
                val xOffset = drawerView.width * slideOffset
                val xOffsetDiff = content!!.width * diffScaledOffset / 2
                val xTranslation = xOffset - xOffsetDiff
                content!!.translationX = xTranslation
            }
        })
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.nav_all_categories -> {
                startActivity(Intent(applicationContext, AllCategories::class.java))
            }
            R.id.nav_favorite -> {
                startActivity(Intent(applicationContext, FavoriteActivity::class.java))
            }
            R.id.nav_login -> {
                startActivity(Intent(applicationContext, PhoneAuthActivity::class.java))
            }
            R.id.nav_home -> {
                startActivity(Intent(applicationContext, UserDashboard::class.java))
            }
            R.id.nav_settings -> {
                startActivity(Intent(applicationContext, SettingsActivity::class.java))
            }
            R.id.nav_chatroom -> {
                startActivity(Intent(applicationContext, ChatActivity::class.java))
            }
            else -> {
                Toast.makeText(this, "Under Development", Toast.LENGTH_SHORT).show()
            }
        }
        return true
    }

    override fun onBackPressed() {
        if (drawerLayout!!.isDrawerVisible(GravityCompat.START)) {
            drawerLayout!!.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    //endregion

    //region featured
    private fun featuredRecycler() {

        featuredRecycler?.setHasFixedSize(true)
        featuredRecycler?.layoutManager = LinearLayoutManager(this, RecyclerView.HORIZONTAL, false)
        //todo we can add array of url images and downloaded with glide or picasso

        val container = findViewById<View>(R.id.shimmerFrameLayout) as ShimmerFrameLayout
        container.startShimmer()

        var list = ArrayList<String>()

        list.add("going-out")
        list.add("natural-geographical")
        list.add("leisure-outdoor")

        Log.d(TAG, ("featuredRecycler: " + this::latitude.isInitialized + this::longitude.isInitialized))
        Log.d(TAG, ("featuredRecycler: $latitude<-->$longitude"))

        if(this::latitude.isInitialized || this::longitude.isInitialized) {
            viewModel.discoverHerePlacesHereDeveloper("dmLgAQo631UJfwF5R2hH", "391hkRjz5Z3Ee1h3wz6Kng",
                    "$latitude,$longitude", list)
        }

        viewModel.discoverHerePlacesHereDeveloperLiveData.observe(this, Observer {
            for (item in it.results.items) {
                var str: StringBuilder
                if (item.title.length > 25) {
                    str = java.lang.StringBuilder(item.title)
                    str.insert(25, "\n").toString()
                } else {
                    str = java.lang.StringBuilder(item.title)
                }

                listOfFeaturedAdapter.add(FeaturedHelperClass(item.icon, str.toString(), item.category.title, item.id))
            }

            if (featuredRecycler?.adapter != null) {
                featuredRecycler?.adapter?.notifyDataSetChanged()
            } else {
                featuredRecycler?.adapter = FeaturedAdapter(listOfFeaturedAdapter, this)
            }

            container.stopShimmer()
            container.visibility = GONE
            featuredRecycler?.visibility = View.VISIBLE

        })

    }


    //endregion

    //region categories
    private fun categoriesRecycler() {
        //All Gradients
        val gradient2 = GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, intArrayOf(R.color.card1, -R.color.card1))
        val gradient1 = GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, intArrayOf(R.color.card2, -R.color.card2))
        val gradient3 = GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, intArrayOf(R.color.card3, -R.color.card3))
        val gradient4 = GradientDrawable(GradientDrawable.Orientation.LEFT_RIGHT, intArrayOf(R.color.card4, -R.color.card4))
        val categoriesHelperClasses = ArrayList<CategoriesHelperClass>()

        categoriesHelperClasses.add(CategoriesHelperClass(gradient1, R.drawable.school_cat, getString(string.education)))
        categoriesHelperClasses.add(CategoriesHelperClass(gradient2, R.drawable.hospital_cat, getString(string.hospital)))
        categoriesHelperClasses.add(CategoriesHelperClass(gradient3, R.drawable.res_cat, getString(string.restaurant)))
        categoriesHelperClasses.add(CategoriesHelperClass(gradient4, R.drawable.shopping_cat, getString(string.shopping)))
        categoriesHelperClasses.add(CategoriesHelperClass(gradient1, R.drawable.transport_cat, getString(string.transport)))

        categoriesRecycler!!.setHasFixedSize(true)
        categoriesAdapter = CategoriesAdapter(categoriesHelperClasses)
        categoriesRecycler!!.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
        categoriesRecycler!!.adapter = categoriesAdapter
    }

    //endregion

    //region mostViewed recycler view
    private fun mostViewedRecycler() {

        mostViewedRecycler!!.setHasFixedSize(true)

        mostViewedRecycler!!.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)

        // app_id dmLgAQo631UJfwF5R2hH //app_code 391hkRjz5Z3Ee1h3wz6Kng

        container = findViewById<View>(R.id.shimmerFrameLayoutMostViewed) as ShimmerFrameLayout

        container.startShimmer()

        var list = ArrayList<String>()

        list.add("coffee-tea")
        list.add("restaurant")
        list.add("snacks-fast-food")

        Log.d(TAG, ("mostViewedRecycler: " + this::latitude.isInitialized + this::longitude.isInitialized))
        Log.d(TAG, ("mostViewedRecycler: $latitude<-->$longitude"))

        if(this::latitude.isInitialized || this::longitude.isInitialized) {
            print(this::latitude.isInitialized || this::longitude.isInitialized)

            viewModel.discoverExplorePlacesHereDeveloper("dmLgAQo631UJfwF5R2hH", "391hkRjz5Z3Ee1h3wz6Kng",
                    "$latitude,$longitude", list)
        }
        viewModel.discoverExplorePlacesHereDeveloperLiveData.observe(this, Observer {
            for (item in it.results.items.asReversed()) {
                var str: StringBuilder
                if (item.title.length > 25) {
                    str = java.lang.StringBuilder(item.title)
                    str.insert(25, "\n").toString()
                } else {
                    str = java.lang.StringBuilder(item.title)
                }
                listOfMostViewedAdapter.clear()
                listOfMostViewedAdapter.add(MostViewedHelperClass(item.icon, str.toString(), item?.alternativeNames?.get(0)?.name
                        ?: "",
                        item.category.title ?: "Category?", item.openingHours?.label
                        ?: "Opening Hours",
                        item.openingHours?.text?.replace("<br/>", "\n") ?: "Not Available?",
                        rating = item.averageRating ?: 4.0
                        ,item.id
                ))
                Log.d(TAG, "context = ${it.search.context.href}")
            }
            if (mostViewedRecycler?.adapter != null) {
                mostViewedRecycler?.adapter?.notifyDataSetChanged()
            } else {
                mostViewedRecycler?.adapter = MostViewedAdapter(listOfMostViewedAdapter, this)
            }
            container.stopShimmer()
            container.visibility = GONE
            mostViewedRecycler?.visibility = View.VISIBLE
        })

    }

    private fun getMostImageUrl(xid: String): String {
        return "https://upload.wikimedia.org/wikipedia/commons/5/5b/Atlantis_The_Palm_hotel_from_the_sea%2C_Palm_Jumeirah%2C_Dubai.jpg"
    }


    //endregion

    //region request permission

    /**
     * Checks the dynamically-controlled permissions and requests missing permissions from end user.
     */
    private fun checkPermissions(){
        val missingPermissions: MutableList<String> = ArrayList()
        // check all required dynamic permissions
        for (permission in REQUIRED_SDK_PERMISSIONS) {
            val result = ContextCompat.checkSelfPermission(this, permission)
            if (result != PackageManager.PERMISSION_GRANTED) {
                missingPermissions.add(permission)
            }
        }
        if (missingPermissions.isNotEmpty()) {
            // request all missing permissions
            val permissions = missingPermissions.toTypedArray()
            ActivityCompat.requestPermissions(this, permissions, REQUEST_CODE_ASK_PERMISSIONS)
        } else {
            val grantResults = IntArray(REQUIRED_SDK_PERMISSIONS.size)
            Arrays.fill(grantResults, PackageManager.PERMISSION_GRANTED)
            onRequestPermissionsResult(REQUEST_CODE_ASK_PERMISSIONS, REQUIRED_SDK_PERMISSIONS, grantResults)
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, @NonNull permissions: Array<String>, @NonNull grantResults: IntArray) {
        when (requestCode) {
            REQUEST_CODE_ASK_PERMISSIONS -> {
                var index = permissions.size - 1
                while (index >= 0) {
                    if (grantResults[index] != PackageManager.PERMISSION_GRANTED) {
                        // exit the app if one permission is not granted
                        Toast.makeText(this, "Required permission '" + permissions[index]
                                + "' not granted, exiting", Toast.LENGTH_LONG).show()
                        finish()
                        return
                    }
                    --index
                }
                // all permissions were granted

                fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this)
                fusedLocationProviderClient.lastLocation.addOnSuccessListener {
                    // Got last known location. In some rare situations this can be null.
                    latitude = it.latitude.toString()
                    longitude = it.longitude.toString()
                    Log.d(TAG, "onRequestPermissionsResult: latitude = " + it.latitude)
                    Log.d(TAG, "onRequestPermissionsResult: longitude = " + it.longitude)

                    if(this::latitude.isInitialized || this::longitude.isInitialized) {
                        featuredRecycler()
                        mostViewedRecycler()
                    }
                }

                initialize(application)
            }

            MY_PERMISSIONS_REQUEST_LOCATION -> {

                // If request is cancelled, the result arrays are empty.
                if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                    // permission was granted, yay! Do the
                    // location-related task you need to do.
                    if (ContextCompat.checkSelfPermission(this,
                                    Manifest.permission.ACCESS_FINE_LOCATION)
                            == PackageManager.PERMISSION_GRANTED) {

                        //Request location updates:

                    }
                } else {

                    // permission denied, boo! Disable the
                    // functionality that depends on this permission.
                }
                return
            }

        }
        Log.d(TAG, "onRequestPermissionsResult: " + " all permissions were granted")



    }

    //endregion




    companion object {
        private const val TAG = "UserDashboard"
        const val END_SCALE = 0.7f
    }

    fun callRetailerScreen(view: View) {
        startActivity(Intent(this, RetailerStartUpScreen::class.java))
    }

}